checker/cost.go

```diff
@@ -0,0 +1,432 @@
+	} else {
+		return c.costCreateMap(e)
+	}
+}
+
+func (c *coster) costCreateMap(e *exprpb.Expr) CostEstimate {
+	mapVal := e.GetStructExpr()
+	var sum CostEstimate
+	for _, ent := range mapVal.GetEntries() {
+		key := ent.GetMapKey()
+		sum = sum.Add(c.cost(key))
+
+		sum = sum.Add(c.cost(ent.Value))
+	}
+	return sum.Add(createMapBaseCost)
+}
+
+func (c *coster) costCreateMessage(e *exprpb.Expr) CostEstimate {
+	msgVal := e.GetStructExpr()
+	var sum CostEstimate
+	for _, ent := range msgVal.GetEntries() {
+		sum = sum.Add(c.cost(ent.Value))
+	}
+	return sum.Add(createMessageBaseCost)
+}
+
+func (c *coster) costComprehension(e *exprpb.Expr) CostEstimate {
+	comp := e.GetComprehensionExpr()
+	var sum CostEstimate
+	sum = sum.Add(c.cost(comp.IterRange))
+	sum = sum.Add(c.cost(comp.AccuInit))
+	accuType := c.getType(comp.AccuInit)
+	rangeType := c.getType(comp.IterRange)
+	var varType *exprpb.Type
+
+	switch kindOf(rangeType) {
+	case kindList:
+		varType = rangeType.GetListType().ElemType
+	case kindMap:
+		// Ranges over the keys.
+		varType = rangeType.GetMapType().KeyType
+	case kindDyn, kindError, kindTypeParam:
+		// Set the range iteration variable to type DYN as well.
+		varType = decls.Dyn
+	default:
+		// not a valid range
+		varType = decls.Error
+	}
+
+	// Create a scope for the comprehension since it has a local accumulation variable.
+	// This scope will contain the accumulation variable used to compute the result.
+	c.env = c.env.enterScope()
+	c.env.Add(decls.NewVar(comp.AccuVar, accuType))
+	// Create a block scope for the loop.
+	c.env = c.env.enterScope()
+	c.env.Add(decls.NewVar(comp.IterVar, varType))
+	// Check the variable references in the condition and step.
+	loopCost := c.cost(comp.LoopCondition)
+	stepCost := c.cost(comp.LoopStep)
+	// Exit the loop's block scope before checking the result.
+	c.env = c.env.exitScope()
+	sum = sum.Add(c.cost(comp.Result))
+	// Exit the comprehension scope.
+	c.env = c.env.exitScope()
+
+	rangeCnt := c.sizeEstimate(expr{t: c.getType(comp.IterRange), expr: comp.IterRange})
+	rangeCost := rangeCnt.MultiplyByCost(stepCost.Add(loopCost))
+	sum = sum.Add(rangeCost)
+
+	return sum
+}
+
+func (c *coster) sizeEstimate(t AstNode) SizeEstimate {
+	if l := t.LiteralSize(); l != nil {
+		return SizeEstimate{Min: *l, Max: *l}
+	}
+	if l := c.estimator.EstimateSize(t); l != nil {
+		return *l
+	}
+	return SizeEstimate{Min: 0, Max: math.MaxUint64}
+}
+
+func (c *coster) functionCost(overloadId string, target *AstNode, args []AstNode) CostEstimate {
+	if est := c.estimator.EstimateCallCost(overloadId, target, args); est != nil {
+		return *est
+	}
+	switch overloadId {
+	// O(n) functions
+	case overloads.StartsWithString, overloads.EndsWithString, overloads.StringToBytes, overloads.BytesToString:
+		if len(args) == 1 {
+			return c.sizeEstimate(args[0]).MultiplyByCostFactor(0.1)
+		}
+	case overloads.InList:
+		// TODO: account for optimization where lists are converted to sets?
```

Comment:
If the `CreateList` is composed entirely of constant values, then the optimization applies, but you'd need to get this information from the `CheckedExpr` args. It's probably fine the leave the cost computation as-is here.